File: hostel_management_system_backend/Services/RoomsService.cs
using AutoMapper;
using hostel_management_system_backend.Exceptions;
using Microsoft.EntityFrameworkCore;

public interface IRoomsService
{
    Task<List<RoomReadDto>> GetAllAsync(CancellationToken cancellationToken);
    Task<RoomReadDto?> GetByIdAsync(Guid id, CancellationToken cancellationToken);
    Task<RoomReadDto> CreateAsync(RoomCreateDto dto, CancellationToken cancellationToken);
    Task<RoomReadDto?> UpdateAsync(Guid id, RoomUpdateDto dto, CancellationToken cancellationToken);
    Task<bool> DeleteAsync(Guid id, CancellationToken cancellationToken);
}

public sealed class RoomsService : IRoomsService
{
    private readonly ICrudRepository<Room> _repo;
    private readonly IMapper _mapper;

    public RoomsService(ICrudRepository<Room> repo, IMapper mapper)
    {
        _repo = repo;
        _mapper = mapper;
    }

    public async Task<List<RoomReadDto>> GetAllAsync(CancellationToken cancellationToken)
    {
        var rooms = await _repo.GetAllAsNoTrackingAsync(cancellationToken);
        return _mapper.Map<List<RoomReadDto>>(rooms);
    }

    public async Task<RoomReadDto?> GetByIdAsync(Guid id, CancellationToken cancellationToken)
    {
        var room = await _repo.GetByIdAsNoTrackingAsync(id, cancellationToken);
        if (room is null)
            throw new NotFoundException("Room not found.");

        return _mapper.Map<RoomReadDto>(room);
    }

    public async Task<RoomReadDto> CreateAsync(RoomCreateDto dto, CancellationToken cancellationToken)
    {
        var entity = _mapper.Map<Room>(dto);
        entity.CreatedAt = DateTime.UtcNow;
        entity.UpdatedAt = null;
        entity.IsDeleted = false;

        await _repo.AddAsync(entity, cancellationToken);
        try
        {
            await _repo.SaveChangesAsync(cancellationToken);
        }
        catch (DbUpdateException ex)
        {
            throw new ConflictException("A room with the same unique fields already exists.", "room_conflict", ex);
        }

        return _mapper.Map<RoomReadDto>(entity);
    }

    public async Task<RoomReadDto?> UpdateAsync(Guid id, RoomUpdateDto dto, CancellationToken cancellationToken)
    {
        var entity = await _repo.GetByIdForUpdateAsync(id, cancellationToken);
        if (entity is null)
            throw new NotFoundException("Room not found.");

        _mapper.Map(dto, entity);
        entity.UpdatedAt = DateTime.UtcNow;
        await _repo.SaveChangesAsync(cancellationToken);
        return _mapper.Map<RoomReadDto>(entity);
    }

    public async Task<bool> DeleteAsync(Guid id, CancellationToken cancellationToken)
    {
        var entity = await _repo.GetByIdForUpdateAsync(id, cancellationToken);
        if (entity is null)
            throw new NotFoundException("Room not found.");

        entity.IsDeleted = true;
        entity.DeletedAt = DateTime.UtcNow;
        entity.UpdatedAt = DateTime.UtcNow;
        await _repo.SaveChangesAsync(cancellationToken);
        return true;
    }
}

File: src/rooms/rooms.service.ts
import { Injectable } from '@nestjs/common';

import {
  AppConflictException,
  AppNotFoundException,
} from '../common/exceptions/app-exception';
import { decimalToNumber } from '../common/utils/prisma.util';
import { PrismaService } from '../prisma/prisma.service';
import { RoomCreateDto, RoomReadDto, RoomUpdateDto } from './dto/rooms.dto';

@Injectable()
export class RoomsService {
  constructor(private readonly prisma: PrismaService) {}

  async getAll(): Promise<RoomReadDto[]> {
    const rooms = await this.prisma.room.findMany({
      where: { isDeleted: false },
      orderBy: { createdAt: 'desc' },
    });

    return rooms.map((room) => this.toReadDto(room));
  }

  async getById(id: string): Promise<RoomReadDto> {
    const room = await this.prisma.room.findFirst({
      where: {
        id,
        isDeleted: false,
      },
    });

    if (!room) {
      throw new AppNotFoundException('Room not found.');
    }

    return this.toReadDto(room);
  }

  async create(dto: RoomCreateDto): Promise<RoomReadDto> {
    try {
      const room = await this.prisma.room.create({
        data: {
          hostelId: dto.hostelId,
          roomType: dto.roomType,
          price: dto.price,
          capacity: dto.capacity,
          isAvailable: dto.isAvailable,
          createdAt: new Date(),
          isDeleted: false,
        },
      });

      return this.toReadDto(room);
    } catch {
      throw new AppConflictException('A room with the same unique fields already exists.', 'room_conflict');
    }
  }

  async update(id: string, dto: RoomUpdateDto): Promise<RoomReadDto> {
    const room = await this.prisma.room.findFirst({
      where: {
        id,
        isDeleted: false,
      },
    });

    if (!room) {
      throw new AppNotFoundException('Room not found.');
    }

    const updated = await this.prisma.room.update({
      where: { id },
      data: {
        roomType: dto.roomType,
        price: dto.price,
        capacity: dto.capacity,
        isAvailable: dto.isAvailable,
        updatedAt: new Date(),
      },
    });

    return this.toReadDto(updated);
  }

  async delete(id: string) {
    const room = await this.prisma.room.findFirst({
      where: {
        id,
        isDeleted: false,
      },
    });

    if (!room) {
      throw new AppNotFoundException('Room not found.');
    }

    await this.prisma.room.update({
      where: { id },
      data: {
        isDeleted: true,
        deletedAt: new Date(),
        updatedAt: new Date(),
      },
    });
  }

  private toReadDto(room: {
    id: string;
    hostelId: string;
    roomType: string;
    price: unknown;
    capacity: number;
    isAvailable: boolean;
    createdAt: Date;
    updatedAt: Date | null;
  }): RoomReadDto {
    return {
      id: room.id,
      hostelId: room.hostelId,
      roomType: room.roomType,
      price: decimalToNumber(room.price as never) ?? 0,
      capacity: room.capacity,
      isAvailable: room.isAvailable,
      createdAt: room.createdAt,
      updatedAt: room.updatedAt,
    };
  }
}
